### ISSUE TITLE
Can't sign in using Safari

### ISSUE THREAD

**Issue by BRadHoc** [michib/nextcloud_twofactor_webauthn#52]
Unlike on Webauthn.io when trying to sign in after registering in the nextcloud settings security tab (which does allow the use of Touch ID on Safari). When trying to sign in, it presents a screen asking for security key only.

**Comment by michib** [comment]
Sorry but i don't support issues that provide no information to find the issue. But in this case i'm also unable to provide support because i don't have mac to reproduce the issue.
Please feel free to provide a PR with the fix.

**Comment by vszakats** [comment]
Similar issue here. Using Safari 15.1. Adding new keys work correctly, but on login, clicking on 'Webauthn Devices' will unconditionally jump back to the `/index.php/login/selectchallenge` page. After enabling the 'Web Authentication Modern' option under 'Develop → Experimental Features', this gets improved by actually going to the `/index.php/login/challenge/twofactor_webauthn` page and idly waiting without the expected prompt, and then the key (or pressing on it) is never actually recognized.

Possibly relevant Web Console message:
"[Warning] User gesture is not detected. To use the WebAuthn API, call 'navigator.credentials.create' within user activated events. (challenge.js, line 13)"

I'm fairly certain that Nextcloud WebAuthn login used to work a few years back (with a different server/add-on/Safari version).

Maybe worthy of note that most other WebAuthn-supporting apps (e.g. github.com) do work correctly in this Safari version, even with 'Web Authentication Modern' _disabled_.

**Comment by michib** [comment]
Yes, webauthn should work on a mac. But someone with a mac has to provide a fix, because i can't reproduce it without one

**Comment by vszakats** [comment]
If you need some specific info, I can try providing it, but I'm not an expert of any of the components involved, so providing a PR is not an option for me at least.

**Comment by vszakats** [comment]
Found these pointers, in the hope they might help finding the root cause and help developing a fix:
- <https://support.yubico.com/hc/en-us/articles/360022004600-No-reaction-when-using-WebAuthn-on-macOS-iOS-and-iPadOS>
- <https://github.com/18F/identity-idp/issues/5132>
related fix/patch: <https://github.com/18F/identity-idp/pull/5162/files>
- <https://github.com/w3c/webauthn/issues/1293>

**Comment by vszakats** [comment]
I understand that buying an Apple device to fix this isn't a viable option.  What I can offer is that if you have a test patch/release or a blind fix in a next release, I can make a test with it (in my particular env) and report back. For certain, at the moment, login cannot be worse than the current state on Apple devices (apparently caused by breaking changes done by Apple in Safari.), so any attempted fix would be welcome and may only improve the situation.

**Comment by michib** [comment]
Ok, thanks for the offer. If i have enough time coming weekend i am planning on releasing a version with better logging. Would be great if you can check what output is shown with this version and maybe i can provide a special version you could try out without me releasing something officially

**Comment by vszakats** [comment]
Sounds great to me, thank you for your efforts!

**Comment by michib** [comment]
There is a new Release 0.2.12. Could you try to repeat the process and check the logs. There should be some more shown during the registration process. Maybe this helps to pinpoint the issue

**Comment by vszakats** [comment]
@michib: Thank you for your update. After upgrading on the server to 0.2.12, the behaviour did not change, meaning: Unconditional redirect back to the auth selection page when clicking on Webauthn on the login page. When enabling 'Webauthn Modern' experimental option in Safari, this is fixed and the browser lands on the Webauth Devices page, but here there is no prompt, and inserting or pressing the key does not have any effect.

The console error is also the same, but with a different line number:
![Screen Shot 2021-11-28 at 15 21 15](https://user-images.githubusercontent.com/1446897/143772158-61650ede-fb45-42ad-bb6f-5d29a5e7047e.png)

The expected browser prompt would look like this:
![Screen Shot 2021-11-28 at 15 27 38](https://user-images.githubusercontent.com/1446897/143772251-64b86c13-47d4-4cf0-9df6-4f7f8ac58c42.png)

One more tidbit of information: The 'Webauthn Modern' setting is probably not something that should be required, enabled, e.g. the `github.com` Webauthn login does break, so this appears to be indeed experimental and not necessary for existing login workflows.

**Comment by michib** [comment]
What are the entries on the nextcloud server log regarding your login attempt?

**Comment by michib** [comment]
Oh think i understand the issue. I try to provide a fix

**Comment by michib** [comment]
[twofactor_webauthn-0.2.13-alpha.0.tar.gz](https://github.com/michib/nextcloud_twofactor_webauthn/files/7614092/twofactor_webauthn-0.2.13-alpha.0.tar.gz)
@vszakats Could you install this version on your nextcloud instance and try webauthn authentication without activating "Webauthn Modern"?

**Comment by vszakats** [comment]
@michib: Updated to 0.2.13-alpha.0, kept 'Webauthn Modern' disabled. With this combination, I can report that now the unconditional redirect is _fixed_, meaning that selecting 'Webauthn login' takes to the 'Webauthn Devices' page as expected. Then, clicking on the button 'Use webauthn device' correctly pops up the Safari prompt and tapping on the key was also recognized.

At this point I got a 'Found no credentials on this device.' popup (with two separate keys that were added already), but this might be a different issue altogether. Then I re-added the key under a different name, retried the login and this time the _key was correctly recognized_. It means that it may be necessary to delete and readd keys in Security settings to make them work. The other key was also fixed by deleting the old entry and re-adding it.

It means that your updates in 0.2.13-alpha.0 _did fix_ all the login issues. Thank you very much for doing this. If you need any further testing, I'm glad to do them.

**Comment by michib** [comment]
Thank you for your feedback, i'm glad this issue could be resolved.
You are right about #44, but i'm not sure about #71. Will have to recheck.
